feat!: add HybridResolver

aaronc opened the pull request:
This provides a quick way to get accessed to a merged protodesc.Resolver by building a `*protoregistry.Files` for all the gogo files and then implementing a resolver on top. The merged registry functionality can then only be used when validation of conflicts is needed.

aaronc commented on proto/test_proto/test.proto:
```diff
@@ -361,67 +361,9 @@ extend DefaultsMessage {
-message MyMessageSet {
-  option message_set_wire_format = true;
-  extensions 100 to max;
-}
```
It was necessary to delete this plus all references to it because message sets are a protov1 feature unsupported by the official protobuf and so registering this file descriptor was failing

mark-rushakoff commented on proto/merge.go:
```diff
@@ -186,7 +185,7 @@ type descriptorProcessor struct {
 // process reads from p.globalFileCh and p.appFileCh, processing each file descriptor as appropriate,
 // and sends the processed file descriptors through p.fdCh for eventual return from mergedFileDescriptors.
-// Any errors during processing are sent to ec.ProcessErrCh,
+//  Any errors during processing are sent to ec.ProcessErrCh,
```
This extra space will probably render the line as code in godoc.

aaronc commented on proto/properties.go:
```diff
@@ -596,13 +603,52 @@ func MessageType(name string) reflect.Type {
 var (
-	protoFiles = make(map[string][]byte) // file name => fileDescriptor
+	protoFiles        = make(map[string][]byte) // file name => fileDescriptor
+	gogoProtoRegistry = new(protoregistry.Files)
```
Maybe this can actually be made public like `*GlobalFiles` and then other checks can be performed by the caller if needed?

aaronc replied:
Exported this as `GogoResolver`. Let me know if you see any issue with that

mark-rushakoff replied:
I think it should be okay exported.

mark-rushakoff commented on proto/merge.go:
```diff
@@ -360,3 +325,52 @@ func mergedFileDescriptors(globalFiles *protoregistry.Files, appFiles map[string
+}
+
+func (r *hybridResolver) RangeFiles(f func(fileDescriptor protoreflect.FileDescriptor) bool) {
+	seen := map[protoreflect.FullName]bool{}
```
This could be pre-sized to `GlobalFiles.NumFiles()`, but it probably won't be in any hot loops, so it should also be fine as-is.

mark-rushakoff reviewed (approved):
I like this approach!

aaronc commented:
@mark-rushakoff `BenchmarkRepresentativeMergedFileDescriptors` isn't building because of the argument type changed. I can gunzip and unmarshal all these descriptors but that will affect the benchmark time. How do you recommend the benchmark be updated to reflect the changes here?

mark-rushakoff commented:
`BenchmarkRepresentativeMergedFileDescriptors` has a call to `b.ResetTimer()` before the benchmark loop, so un-gzipping them before that call seems fine, and it is probably less effort than doing a new export from simd.

I'm okay with a tag retraction in go.mod, as the recent tag is only a couple weeks old and should only have been in use in the SDK.

aaronc commented:
What do you think is the best way to approach the retraction? Delete the tag and create a new one? Or do we actually need a retract directive in the go.mod?

aaronc commented:
Also, really confused by the CI errors here. Seems like there's a lot of stuff failing unrelated to any changes in this PR.

mark-rushakoff commented:
I think just a retract directive in go.mod is fine. That might be a slight misuse of retraction, but I think it's okay given the narrow scope here -- "don't use this tag as it was a breaking API change, which we undid in the following tag".

I understand the CI error message -- we are enforcing the leading slash somewhere as an expectation against the cosmos protos we produce. But I am not immediately seeing why that just started happening on this PR.

aaronc commented:
I added a retract directive for 1.4.8

The test failures seem due to registration errors in the `*protoregistry.Files` or `CheckImportPath`. I fixed them by:
* converting the registration errors to warnings
* deleting one test case (`typedeclimport`) which had what I consider to be pathological imports which we shouldn't try to support

Going to move forward with merging this unless anyone sees any issues with this.

julienrbrt reviewed (approved):
Can we get the changelog ready as well?

aaronc commented:
See #63 for the CHANGELOG
